File: be/SpringBoot/src/main/java/com/common/controller/UserController.java
package com.common.controller;

import com.common.dto.response.UserDTO;
import com.common.model.enumerate.EntityStatus;
import com.common.model.sql.User;
import com.common.service.contract.IUserService;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * Controller quản lý User Entity.
 *
 * @created_at 2026-07-31
 * @author txhoan
 */
@RestController
@RequestMapping("${app.api.users-prefix:${server.servlet.context-path:/api/v1}/users}")
public class UserController extends BaseController<User, Long, UserDTO, UserDTO, IUserService> {

    public UserController(IUserService service) {
        super(service);
    }

    @Override
    protected UserDTO toResponse(User entity) {
        if (entity == null) return null;
        return UserDTO.builder()
                .id(entity.getId())
                .username(entity.getUsername())
                .email(entity.getEmail())
                .status(entity.getStatus())
                .tenantId(entity.getTenant() != null ? entity.getTenant().getId() : null)
                .tenantCode(entity.getTenant() != null ? entity.getTenant().getCode() : null)
                .roles(entity.getRoles())
                .createdAt(entity.getCreatedAt())
                .modifiedAt(entity.getModifiedAt())
                .build();
    }

    @Override
    protected User toEntity(UserDTO request) {
        if (request == null) return null;
        return User.builder()
                .username(request.getUsername())
                .email(request.getEmail())
                .status(request.getStatus() != null ? request.getStatus() : EntityStatus.ACTIVE)
                .roles(request.getRoles())
                .build();
    }

    @Override
    protected void updateEntity(UserDTO request, User entity) {
        if (request.getEmail() != null) entity.setEmail(request.getEmail());
        if (request.getStatus() != null) entity.setStatus(request.getStatus());
        if (request.getRoles() != null) entity.setRoles(request.getRoles());
    }
}

File: be/.Net/Swe.API/Controllers/UserController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Swe.BL.DTO.Response;
using Swe.BL.Service;
using Swe.Common.Model;

namespace Swe.API.Controllers;

/// <summary>
/// Controller quản lý User (Tài khoản người dùng).
/// </summary>
/// <created_at>2026-08-03</created_at>
/// <author>txhoan</author>
[Authorize]
[ApiController]
[Route("api/v1/users")]
public class UserController(UserService userService, UserRepository userRepository, TenantRepository tenantRepository) : ControllerBase
{
    [HttpGet("{id:long}")]
    public async Task<IActionResult> GetById([FromRoute] long id)
    {
        try
        {
            var user = await userService.GetByIdAsync(id);
            if (user == null)
            {
                return NotFound(new WrapperApiResponse(404, "Không tìm thấy người dùng", null));
            }

            var roles = await userRepository.GetRolesAsync(user.Id);
            var tenant = user.TenantId.HasValue ? await tenantRepository.GetByIdAsync(user.TenantId.Value) : null;

            var dto = new UserDTO
            {
                Id = user.Id,
                Username = user.Username,
                Email = user.Email,
                Status = user.Status,
                TenantId = user.TenantId,
                TenantCode = tenant?.Code,
                Roles = roles,
                CreatedAt = user.CreatedAt,
                ModifiedAt = user.ModifiedAt
            };

            return Ok(new WrapperApiResponse(200, "Lấy thông tin người dùng thành công", dto));
        }
        catch (Exception ex)
        {
            return StatusCode(500, new WrapperApiResponse(500, "Đã xảy ra lỗi hệ thống: " + ex.Message, null));
        }
    }
}
